src/session_pool/session_pool.js

```diff
@@ -0,0 +1,314 @@
+        checkParamOrThrow(maxSessionAgeSecs, 'options.maxSessionAgeSecs', 'Number');
+        checkParamOrThrow(maxSessionUsageCount, 'options.maxSessionUsageCount', 'Number');
+        checkParamOrThrow(persistStateKeyValueStoreId, 'options.persistStateKeyValueStoreId', 'Maybe String');
+        checkParamOrThrow(persistStateKey, 'options.persistStateKey', 'String');
+        checkParamOrThrow(createSessionFunction, 'options.createSessionFunction', 'Maybe Function');
+
+        // Pool Configuration
+        this.maxPoolSize = maxPoolSize;
+        this.createSessionFunction = createSessionFunction || this._defaultCreateSessionFunction;
+
+        // Session configuration
+        // @TODO: Maybe options.sessionOptions / this.sessionOptions?
+        this.maxSessionAgeSecs = maxSessionAgeSecs;
+        this.maxSessionUsageCount = maxSessionUsageCount;
+
+        // Session keyValueStore
+        this.persistStateKeyValueStoreId = persistStateKeyValueStoreId;
+        this.persistStateKey = persistStateKey;
+
+        // Operative states
+        this.keyValueStore = null;
+        this.sessions = [];
+    }
+
+    /**
+     * Gets count of usable sessions in the pool.
+     * @return {number}
+     */
+    get usableSessionsCount() {
+        return this.sessions.filter(session => session.isUsable()).length;
+    }
+
+    /**
+     * Gets count of blocked sessions in the pool.
+     * @return {number}
+     */
+    get retiredSessionsCount() {
+        return this.sessions.filter(session => !session.isUsable()).length;
+    }
+
+    /**
+     * Starts periodic state persistence and potentially loads SessionPool state from {@link KeyValueStore}.
+     * This function must be called before you can start using the instance in a meaningful way.
+     *
+     * @return {Promise<void>}
+     */
+    async initialize() {
+        this.keyValueStore = await openKeyValueStore(this.persistStateKeyValueStoreId);
+
+        // in case of migration happened and SessionPool state should be restored from the keyValueStore.
+        await this._maybeLoadSessionPool();
+
+        events.on(ACTOR_EVENT_NAMES_EX.PERSIST_STATE, this.persistState.bind(this));
+    }
+
+    /**
+     * Gets session.
+     * If there is space for new session, it creates and return new session.
+     * If the session pool is full, it picks a session from the pool,
+     * If the picked session is usable it is returned, otherwise it creates and returns a new one.
+     *
+     * @return {Promise<Session>}
+     */
+    async retrieveSession() {
+        // If we have enough space for session. Return newly created session.
+        if (this._hasSpaceForSession()) {
+            return this._createSession();
+        }
+
+        // For example that developer can plug different picking algorithms such as the Lukášův
+        // Maybe a should have pickSession function to be customizable.
+        const pickedSession = this._pickSession();
```

Comment:
I think that we should scratch different picking algorithms for the time being and remove the comment.